src/main/java/org/folio/fqm/config/EntityTypePermissionsAspect.java

```diff
@@ -0,0 +1,146 @@
+
+  private final Map<MethodSignature, Integer> indexCache = new HashMap<>();
+
+  /**
+   * Handle methods that accept an entity type or query ID
+   */
+  @Around("@annotation(org.folio.fqm.annotation.EntityTypePermissionsRequired) && execution(* *(.., java.util.UUID, ..))")
+  private Object validatePermissionsWithId(ProceedingJoinPoint joinPoint) throws Throwable {
+    MethodSignature methodSignature = ((MethodSignature) joinPoint.getSignature());
+    EntityTypePermissionsRequired annotation = methodSignature.getMethod().getAnnotation(EntityTypePermissionsRequired.class);
+
+    return validatePermissions(joinPoint, (UUID id) -> {
+      var entityTypeId = switch (annotation.idType()) {
+        case ENTITY_TYPE -> id;
+        case QUERY -> queryRepository.getQuery(id, true)
+          .orElseThrow(() -> new EntityTypeNotFoundException(id))
+          .entityTypeId();
+      };
+      return getEntityTypeFromId(entityTypeId);
+    });
+  }
+
+  /**
+   * Handle methods that accept an EntityType object
+   */
+  @Around("@annotation(org.folio.fqm.annotation.EntityTypePermissionsRequired) && execution(* *(.., org.folio.querytool.domain.dto.EntityType, ..))")
+  private Object validatePermissionsWithEntityType(ProceedingJoinPoint joinPoint) throws Throwable {
+    return validatePermissions(joinPoint, Function.identity());
+  }
+
+  /**
+   * Handle methods that accept a SubmitQuery object
+   */
+  @Around("@annotation(org.folio.fqm.annotation.EntityTypePermissionsRequired) && execution(* *(.., org.folio.querytool.domain.dto.SubmitQuery, ..))")
+  private Object validatePermissionsWithSubmitQuery(ProceedingJoinPoint joinPoint) throws Throwable {
+    return validatePermissions(joinPoint, (SubmitQuery query) -> getEntityTypeFromId(query.getEntityTypeId()));
+  }
+
+  /**
+   * Handle methods that accept a ContentsRequest object
+   */
+  @Around("@annotation(org.folio.fqm.annotation.EntityTypePermissionsRequired) && execution(* *(.., org.folio.querytool.domain.dto.ContentsRequest, ..))")
+  private Object validatePermissionsWithContentsRequest(ProceedingJoinPoint joinPoint) throws Throwable {
+    return validatePermissions(joinPoint, (ContentsRequest request) -> getEntityTypeFromId(request.getEntityTypeId()));
+  }
+
+  private EntityType getEntityTypeFromId(UUID entityTypeId) {
+    return entityTypeRepository.getEntityTypeDefinition(entityTypeId)
+      .orElseThrow(() -> new EntityTypeNotFoundException(entityTypeId));
+  }
+
+  /**
+   * Validates that the user has the necessary permissions to perform the operation.
+   *
+   * @param entityTypeConverter - A function to retrieve an EntityType object from the target method parameter.
+   */
+  @SuppressWarnings("unchecked") // If the types are wrong, we want to fail loudly
+  private <T> Object validatePermissions(ProceedingJoinPoint joinPoint, Function<T, EntityType> entityTypeConverter) throws Throwable {
+    // 1. Retrieve the annotated method and annotation details.
```

Comment:
There's a lot going on in this method, and breaking it up more didn't seem to help much, so I opted for numbered comments so that it reads like, "step 1: do A. Step 2: Do B. Step 3..."